### DISCUSSION TITLE
Information request for the custom smooth bitmap font format

### DISCUSSION THREAD

**Discussion by ChaoticNeutralCzech** [Bodmer/TFT_eSPI#2714] [Q&A - Fonts]
**TL;DR**: Need to find documentation of the font format used in file [Tools/Create_Smooth_Font/Create_font/FontFiles/Final-Frontier28.h](https://github.com/Bodmer/TFT_eSPI/blob/master/Tools/Create_Smooth_Font/Create_font/FontFiles/Final-Frontier28.h) so I can create my own smooth font pixel-by-pixel rather than using [your Processing sketch](https://github.com/Bodmer/TFT_eSPI/blob/master/Tools/Create_Smooth_Font/Create_font/Create_font.pde) to convert a vector font (TTF/OTF) into this format

Hi,  
I am trying to build a miniature of [this railway station departure board](https://github.com/ChaoticNeutralCzech/myRandomImages/blob/main/ytyGAx.jpg?raw=true), which involves displaying custom bitmap fonts on a ST7789 LCD. That is, fonts I (re)drew bit-by-bit, not conversions of vector fonts like the [online tool](http://oleddisplay.squix.ch/) or [Processing sketch](https://github.com/Bodmer/TFT_eSPI/blob/master/Tools/Create_Smooth_Font/Create_font/Create_font.pde) does. I already reverse-engineered the GFXFF format (used by fonts like [Fonts/Custom/Roboto_Thin_24.h](https://github.com/Bodmer/TFT_eSPI/blob/master/Fonts/Custom/Roboto_Thin_24.h)) and made [a Python script](https://github.com/ChaoticNeutralCzech/bdf2tft) to convert FontForge-exported BDF fonts into it. 

Anyway, there is a high-segment LCD clock at the top of the board that I would like to replicate. Its high-res (18pt) approximation (see the first row in image) is sufficiently accurate but I need to scale it down to represent it at half the resolution. There are several ways I could go about this:  
![Three ways how to scale down the font: no AA, simple AA and subpixel AA.](https://raw.githubusercontent.com/ChaoticNeutralCzech/myRandomImages/main/GeascriptRenderingExamples.png)  
Obviously, the best one is the subpixel-anti-aliased version (technique known as ClearType), with the simply-anti-aliased version not far behind. I can use my Python script to make the 1bpp font but that looks objectively awful. (Its advantage are sharp characters `1` and `:` but I think I will shift those by half a pixel in the final version anyway so they will be just as sharp.)

To create the simply-anti-aliased version, I could write my own script that would create a font like [Tools/Create_Smooth_Font/Create_font/FontFiles/Final-Frontier28.h](https://github.com/Bodmer/TFT_eSPI/blob/master/Tools/Create_Smooth_Font/Create_font/FontFiles/Final-Frontier28.h). Currently, that requires me to learn Processing to reverse-engineer the [generator sketch](https://github.com/Bodmer/TFT_eSPI/blob/master/Tools/Create_Smooth_Font/Create_font/Create_font.pde) for this format because its output is not obvious (probably features compression such as RLE). (I already tried creating a vector font from my bitmap and feed it to the sketch but failed.) To save me the trouble of reverse-engineering, could somebody please describe how the VLW format and more importantly the accompanying header file is structured?

Anyway, you may think that the ClearType-like subpixel-anti-aliasing is not possible with your sketch but I think I could use a 4-bit-indexed-color sprite with a grayscale palette and render the font into it, and then swap the 16 colors for the ones required for subpixel manipulation (only 8 are actually needed). (This also allows support for 180° rotation by swapping the R and G channels in the palette, as well as reasonably easy adjustment of BG & FG colors by recalculating the palette.)

Either way, I believe that I could make the font look pretty good at small resolution if I can semi-manually create files in the format.

Also, you may say "it's just 12 characters, why bother with a font and not use a set of 4-bit-indexed-color bitmaps?" but a font makes the ESP32 code simpler, and also will come in handy when I try to recreate high-res departure boards: no widely available display has sufficient pixel density to show [this board](https://www.vlakemjednoduse.cz/wp-content/uploads/2018/12/Breclav-2023-05_21_small.jpg) (very common at Czech stations, its resolution is exactly 2x the previous one with the same dimensions) at the required approx. 1:87 scale but the 320x172 1.47" IPS is exactly up to the task if subpixel-anti-aliasing effectively doubles the pixel density. Drawing proportional UTF-8 text with a font featuring over 128 characters using individual bitmaps would be ridiculous. About as ridiculous as creating 7 (!) 1bpp fonts, one for each of the foreground colors, and printing them over each other.

**Comment by ChaoticNeutralCzech** [comment]
(By the way, the high-segment font is a variation of [*Geascript 88*](https://www.bhb.pt/en/images/Produtos/Energia/Sistemas_Informacao/Paineis_LCD/Segmentos/BROCHURE_GEAVISION_SEGMENT_VF.pdf#page=8) by *Elektročas* or perhaps *Mobatime*. The more notorious [*Geascript 38*](https://www.bhb.pt/en/images/Produtos/Energia/Sistemas_Informacao/Paineis_LCD/Segmentos/BROCHURE_GEAVISION_SEGMENT_VF.pdf#page=7) is also used at Czech stations like [Břeclav](https://www.vlakemjednoduse.cz/wp-content/uploads/2018/12/Breclav-2023-05_21_small.jpg).) I will publish my source code and schematic when I finish the [MVP](https://en.wikipedia.org/wiki/Minimum_viable_product).
